### PR TITLE
#88: Rename Foundation Core Java module with latest aiSSEMBLE landscape

### PR DESCRIPTION
**Pull request by habibimoiz**
#88 Rename Foundation Core Java module with latest aiSSEMBLE landscape:
  - Renamed the Foundation Core Java package
  - Renamed the Foundation Metadata Core package.
  - Renamed the Foundation Metadata Core Producer Java package.
  - Renamed all the child directories, classes, and methods.
  - Corrected all the imports inside of the velocity templates to ensure a successful build of the Foundation Core Java, Metadata Core and Metadata Core Producer modules.
  - Create Baton migrations for project updates ensuring the change does not break during project upgrades.

### CODE REVIEW

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.boozallen.aissemble.upgrade.util.FileUtils.replaceLiteralInFile;
+
+public class FoundationCoreJavaAiopsReferencesMigration extends AbstractAissembleMigration {
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java:26]: Q: Can we not re-use/generalize the migration we already have for updating Java package names?

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    public static final Logger logger = LoggerFactory.getLogger(FoundationCoreJavaAiopsReferencesMigration.class);
+    public static List<String> VERSION_AISSEMBLE_PRE_1_7_0 = new ArrayList<>();
+    public static List<String> VERSION_AISSEMBLE_POST_1_7_0 = new ArrayList<>();
+
+    /**
+     * Function to add packages to the VERSION_AISSEMBLE_PRE_1_7_0 & VERSION_AISSEMBLE_POST_1_7_0 list.
+     * @param n/a.
+     * @return n/a.
+     */
+    public static void addPackagesToList(){
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java:37]: A: I think we might end up calling this multiple times.  Might just switch to using a static initialization that's common for these types of constant sets. Something like:
```
ublic static List<String> VERSION_AISSEMBLE_POST_1_7_0 = Collections.unmodifiableList(
   "com.boozallen.aips.core.cdi",
   ...,
   "com.boozallen.aiops.core.filestore"
)
```

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+     * Performs the migration if the shouldExecuteOnFile() returns true.
+     * @param file file to migrate
+     * @return isMigrated - Whether the file was migrated successfully.
+     */
+    @Override
+    protected boolean performMigration(File file) {
+        boolean isMigrated = false;
+
+        try {
+            for (int i=0; i < VERSION_AISSEMBLE_PRE_1_7_0.size(); i++) {
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java:84]: S: Trying to coordinate to Lists by index is very susceptible to bugs.  We should either use a `Map<String,String>` or create an class like `PackageUpdate` that has the old name and new name then use a `List<PackageUpdate>`

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import java.io.FileReader;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.boozallen.aissemble.upgrade.util.FileUtils.replaceLiteralInFile;
+
+public class FoundationCoreJavaAiopsReferencesMigration extends AbstractAissembleMigration {
+
+    public static final Logger logger = LoggerFactory.getLogger(FoundationCoreJavaAiopsReferencesMigration.class);
+    public static List<String> VERSION_AISSEMBLE_PRE_1_7_0 = new ArrayList<>();
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java:29]: A: Let's use a more descriptive name. What _is_ this?

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    @Override
+    protected boolean shouldExecuteOnFile(File file) {
+        boolean shouldExecute = false;
+        addPackagesToList();
+
+        try (BufferedReader foundationCoreJavaConfig = new BufferedReader((new FileReader(file)))) {
+
+            String line;
+            while((line = foundationCoreJavaConfig.readLine()) !=null && !shouldExecute) {
+                for (int i=0; i < VERSION_AISSEMBLE_PRE_1_7_0.size(); i++) {
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java:61]: A: Should be pretty easy and a bit more efficient to use the concatenate the list of literals into a regex `Pattern`.
```java
String regex = oldPackageNames.stream()
   .map(Pattern::quote)
   .collect(Collectors.joining("|"))
Pattern search = Pattern.compile(regex);
```

At that point you can probably reuse some of our util logic.

File: foundation/foundation-upgrade/src/test/resources/specifications/v1_7_0/foundation-core-java-migration.feature [hunk 6]
```diff
@@ -0,0 +1,26 @@
+Feature: Foundation Core Java upgrade migration
+
+  Scenario: Upgrade Foundation Core Java upgrade migration
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/test/resources/specifications/v1_7_0/foundation-core-java-migration.feature:3]: A: Scenario titles all being the same is odd. Each scenario should be titled such that it clearly conveys what specific case is being covered by the test.

File: foundation/foundation-upgrade/src/test/resources/specifications/v1_7_0/foundation-core-java-migration.feature [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    Given a Java file references the cdi package
+    When the 1.7.0 foundation core java package migration executes
+    Then the references are updated to com.boozallen.aissemble.cdi
+
+  Scenario: Upgrade Foundation Core Java upgrade migration
+    Given a Java file references the metadata package
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/test/resources/specifications/v1_7_0/foundation-core-java-migration.feature:9]: A: if all of these are the same except the package name is changing, it might make sense to convert these to a single Scenario Outline.

**Review by ewilkins-csi — APPROVED** [review]

File: extensions/extensions-data-delivery/extensions-data-delivery-spark/src/test/java/com/boozallen/aiops/data/delivery/TestCdiContext.java [hunk 8]
```diff
@@ -10,18 +10,15 @@
  * #L%
  */
 
-import com.boozallen.aiops.core.cdi.CdiContext;
+import com.boozallen.aissemble.core.cdi.CdiContext;
 import com.boozallen.aiops.data.delivery.messaging.PipelineMessagingConfig;
 import com.boozallen.aiops.data.delivery.messaging.pipeline.steps.InOnlyTestStep;
```
Comment (aaron-gary) [inline extensions/extensions-data-delivery/extensions-data-delivery-spark/src/test/java/com/boozallen/aiops/data/delivery/TestCdiContext.java:15]: Q: Is data delivery going to be processed in another PR?  i.e. aiops removed

File: foundation/foundation-upgrade/src/test/resources/test-files/v1_7_0/FoundationCoreJavaMigration/migration/Metadata.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+ * #L%
+ */
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.jboss.weld.environment.se.WeldContainer;
+
+import com.boozallen.aiops.core.metadata.MetadataAPI;
```
Comment (aaron-gary) [inline foundation/foundation-upgrade/src/test/resources/test-files/v1_7_0/FoundationCoreJavaMigration/migration/Metadata.java:19]: Change to aissemble.core(?)

**Review by aaron-gary — APPROVED** [review]

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java (hunk 1 above)
Comment (habibimoiz) [reply foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java]: This was a great idea! I just merged both of the classes into one called "AiopsReferenceMigration".

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java (hunk 4 above)
Comment (habibimoiz) [reply foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java]: Updated the name

File: foundation/foundation-upgrade/src/test/resources/specifications/v1_7_0/foundation-core-java-migration.feature (hunk 6 above)
Comment (habibimoiz) [reply foundation/foundation-upgrade/src/test/resources/specifications/v1_7_0/foundation-core-java-migration.feature]: Updated the titles in the new migration feature file.

File: foundation/foundation-upgrade/src/test/resources/specifications/v1_7_0/foundation-core-java-migration.feature (hunk 7 above)
Comment (habibimoiz) [reply foundation/foundation-upgrade/src/test/resources/specifications/v1_7_0/foundation-core-java-migration.feature]: Changed to a single scenario and file.

File: foundation/foundation-upgrade/src/test/resources/test-files/v1_7_0/FoundationCoreJavaMigration/migration/CdiContainerFactory.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+ * #L%
+ */
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.jboss.weld.environment.se.WeldContainer;
+
+import com.boozallen.aiops.core.cdi.CdiContainer;
```
[the remark this answers is not in the document]
Comment (habibimoiz) [reply foundation/foundation-upgrade/src/test/resources/test-files/v1_7_0/FoundationCoreJavaMigration/migration/CdiContainerFactory.java]: These files have been removed since we moved to a single scenario to cover these test cases.

File: foundation/foundation-upgrade/src/test/resources/test-files/v1_7_0/FoundationCoreJavaMigration/migration/Metadata.java (hunk 9 above)
Comment (habibimoiz) [reply foundation/foundation-upgrade/src/test/resources/test-files/v1_7_0/FoundationCoreJavaMigration/migration/Metadata.java]: These files have been removed since we moved to a single scenario to cover these test cases.

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java (hunk 3 above)
Comment (habibimoiz) [reply foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_7_0/FoundationCoreJavaAiopsReferencesMigration.java]: Changed to a Map.
